### DISCUSSION TITLE
pyo3 and inspect.get_annotations

### DISCUSSION THREAD

**Discussion by dean-shaff** [PyO3/pyo3#3054] [Questions] [answered]
I would like to be able to get type information about my Rust structs at runtime. 

```rust
#[pyclass]
#[pyo3(text_signature = "(start, end)")]
pub struct Foo {
    #[pyo3(get)]
    pub start: i64,
    #[pyo3(get)]
    pub end: i64,
}
```

in Python land: 

```python
>>> import inspect
>>> from my_module import Foo 
>>> inspect.get_annotations(Foo)
{}
```

Ideally, I would like to see something like 

```python
>>> inspect.get_annotations(Foo)
{"start": <class 'int'>, "end": <class 'int'>}
```

Looking at the `get_annotations` [source code](https://github.com/python/cpython/blob/main/Lib/inspect.py#L174) I see that for classes, we're calling `cls.__dict__["__annotations__"]`. I've tried setting `Span`'s `__dict__` attribute (I know this is a bit of a hack!), but I get a RuntimeError at import time. How might I go about supporting this behaviour?

**Comment by konstin** [comment]
This looks like the same problem as https://github.com/PyO3/pyo3/issues/510. To summarize, python unfortunately simply doesn't allow you to set `__annotations__` from native code.

**Comment by dean-shaff** [comment, marked as answer]
I implemented a work around for this.

```rust
#[pyclass]
#[pyo3(text_signature = "(start, end)")]
pub struct Foo {
    #[pyo3(get)]
    pub start: i64,
    #[pyo3(get)]
    pub end: i64,
}

#[pymethods]
impl Foo {
    #[classattr]
    pub fn __annotations__(py: Python<'_>) -> PyResult<Py<PyAny>> 
    {
        let builtins = PyModule::import(py, "builtins")?;
        let t = builtins.getattr(intern!(py, "int"))?;
        let dict = PyDict::new(py);
        dict.set_item("start", t)?; 
        dict.set_item("end", t)?;
        Ok(dict.into()) 
    }
}
```

Now in Python land

```python
>>> import inspect
>>> from my_module import Foo 
>>> inspect.get_annotations(Foo)
{"start": <class 'int'>, "end": <class 'int'>}
```

**Reply by konstin** [reply]
oh, i never even considered cpython would just accept a class attribute!

for you code, i think you can skip the import by using [PyInt](https://docs.rs/pyo3/latest/pyo3/types/struct.PyInt.html#) directly.

**Comment by kngwyu** [comment]
Interesting. Automatically generating this class attr would be an OK approach if we can't find any alternative.
